util/named_requirement_verification/legacy_bidirectional_iterator.h

```diff
@@ -0,0 +1,145 @@
+/*******************************************************************************
+//
+//  SYCL 2020 Conformance Test Suite
+//
+//  Provides class to verify conformity with named requirement
+//  LegacyBidirectionalIterator
+//
+*******************************************************************************/
+
+#ifndef __SYCLCTS_TESTS_ITERATOR_REQUIREMENTS_LEGACY_BIDIRECTIONAL_ITERATOR_H
+#define __SYCLCTS_TESTS_ITERATOR_REQUIREMENTS_LEGACY_BIDIRECTIONAL_ITERATOR_H
+
+#include "common.h"
+#include "legacy_forward_iterator.h"
+
+namespace named_requirement_verification {
+
+/**
+ * @brief Class helps to verify conformity to LegacyBidirectionalIterator named
+ * requirement. Safe to use inside kernel with SYCL 2020.
+ *
+ */
+class legacy_bidirectional_iterator_requirement {
+ public:
+  // Will be used as size of container for error messages
+  // Value should be equal to the number of add_error invocations
+  // Don't forget to update this value if there is any changes in class
+  // As we also verify other requirements, we have to keep in mind that result
+  // of those verifications also increase size of container with messages
+  static constexpr int count_of_possible_errors =
+      legacy_forward_iterator_requirement::count_of_possible_errors + 8;
+
+ private:
+  error_messages_container<count_of_possible_errors> m_errors;
+
+ public:
+  /**
+   * @brief Member function preform different checks for the requirement
+   * verification
+   *
+   * @tparam It Type of iterator for verification
+   * @return std::pair<bool,array<string_view>> First represents
+   * satisfaction of the requirement. Second contains error messages
+   */
+  template <typename It>
+  auto is_satisfied_for(It valid_iterator, const size_t container_size) {
+    auto legacy_forward_iterator_res =
+        legacy_forward_iterator_requirement{}.is_satisfied_for<It>(
+            valid_iterator, container_size);
+
+    if (legacy_forward_iterator_res.first == false) {
+      m_errors.add_errors(legacy_forward_iterator_res.second);
+    }
+
+    using it_traits = std::iterator_traits<It>;
+
+    constexpr bool is_dereferenceable = is_dereferenceable_v<It>;
+    constexpr bool can_pre_increment =
+        type_traits::has_arithmetic::pre_increment_v<It>;
+    constexpr bool can_post_increment =
+        type_traits::has_arithmetic::post_increment_v<It>;
+    constexpr bool can_pre_decrement =
+        type_traits::has_arithmetic::pre_decrement_v<It>;
+    constexpr bool can_post_decrement =
+        type_traits::has_arithmetic::post_decrement_v<It>;
+    constexpr bool has_reference_member =
+        type_traits::has_field::reference_v<It>;
+    constexpr bool has_value_type_member =
+        type_traits::has_field::value_type_v<It>;
+
+    if constexpr (can_pre_increment && can_pre_decrement) {
+      if (std::is_same_v<decltype(--(++std::declval<It&>())), It&> == false) {

```

Comment:
```suggestion
      if (!std::is_same_v<decltype(--(++std::declval<It&>())), It&>) {

```

Reply:
This comment has been applied [here](https://github.com/KhronosGroup/SYCL-CTS/pull/354/commits/6790e9d80b7ed42d557b57974467d2bc7c07a4c6#diff-98e2bb40cc64308f2138f8199c7e0eaa1fc5eb582f99d3192586afe30c8fd2d0R72) and in other cases